pkg/router/template/util/util.go

```diff
@@ -94,3 +96,176 @@ func GenerateBackendNamePrefix(termination routev1.TLSTerminationType) string {
 
 	return prefix
 }
+
+// ParseDurationWithErrors is time.ParseDuration, modified with meaningful
+// error messages for invalid input. It parses a duration string.
+// A duration string is a possibly signed sequence of
+// decimal numbers, each with optional fraction and a unit suffix,
+// such as "300ms", "-1.5h" or "2h45m".
+// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+// Most of the comments are from the original time.ParseDuration.
+func ParseDurationWithErrors(s string) (time.Duration, error) {
+	// [-+]?([0-9]*(\.[0-9]*)?[a-z]+)+
+	orig := s
+	var d uint64
+	neg := false
+
+	// Consume [-+]?
+	if s != "" {
+		c := s[0]
+		if c == '-' || c == '+' {
+			neg = c == '-'
+			s = s[1:]
+		}
+	}
+	// Special case: if all that is left is "0", this is zero.
+	if s == "0" {
+		return 0, nil
+	}
+	if s == "" {
+		return 0, errors.New("time: invalid duration, contains only sign/s " + orig)
+	}
+	for s != "" {
+		var (
+			v, f  uint64      // integers before, after decimal point
+			scale float64 = 1 // value = v + f/scale
+		)
+
+		var err error
+
+		// The next character must be [0-9.]
+		if !(s[0] == '.' || '0' <= s[0] && s[0] <= '9') {
+			return 0, errors.New("time: invalid duration, not a number " + orig)
+		}
+		// Consume [0-9]*
+		pl := len(s)
+		v, s, err = leadingInt(s)
+		if err != nil {
+			return 0, errors.New(err.Error() + orig)
+		}
+		pre := pl != len(s) // whether we consumed anything before a period
+
+		// Consume (\.[0-9]*)?
+		post := false
+		if s != "" && s[0] == '.' {
+			s = s[1:]
+			pl := len(s)
+			f, scale, s = leadingFraction(s)
+			post = pl != len(s)
+		}
+		if !pre && !post {
+			// no digits (e.g. ".s" or "-.s")
+			return 0, errors.New("time: invalid duration, not a number " + orig)
+		}
+
+		// Consume unit.
+		i := 0
+		for ; i < len(s); i++ {
+			c := s[i]
+			if c == '.' || '0' <= c && c <= '9' {
+				break
+			}
+		}
+		if i == 0 {
+			return 0, errors.New("time: missing unit in duration " + orig)
+		}
+		u := s[:i]
+		s = s[i:]
+		unit, ok := UnitMap[u]
+		if !ok {
+			return 0, errors.New("time: unknown unit " + u + " in duration " + orig)
+		}
+		if v > 1<<63/unit {
+			// overflow
+			return 0, errors.New("time: invalid duration, overflow " + orig)
+		}
+		v *= unit
+		if f > 0 {
+			// float64 is needed to be nanosecond accurate for fractions of hours.
+			// v >= 0 && (f*unit/scale) <= 3.6e+12 (ns/h, h is the largest unit)
+			v += uint64(float64(f) * (float64(unit) / scale))
+			if v > 1<<63 {
+				// overflow
+				return 0, errors.New("time: invalid duration, overflow " + orig)
+			}
+		}
+		d += v
+		if d > 1<<63 {
+			return 0, errors.New("time: invalid duration, overflow " + orig)
+		}
+	}
+	if neg {
+		return -time.Duration(d), nil
+	}
+	if d > 1<<63-1 {
+		return 0, errors.New("time: invalid duration, overflow " + orig)
```

Comment:
Any reason not simply to return `haproxyMaxTimeout` in case of overflow?  

If you prefer to return an error value here and handle overflow in `clipHAProxyTimeoutValue`, consider using a typed error value, something like this:
```go
// OverflowError represents an overflow error from ParseDurationWithErrors.
type OverflowError struct {
	error
}

func ParseDurationWithErrors(s string) (time.Duration, error) {
	// ...
	return 0, OverflowError{errors.New("time: invalid duration, overflow " + orig)}
	// ...
}

func clipHAProxyTimeoutValue(val string) string {
	// ...
	duration, err := templateutil.ParseDurationWithErrors(val)
	switch err.(type) {
	case nil:
	case templateutil.OverflowError:
		return haproxyMaxTimeout
	default:
		return haproxyDefaultTimeout
	}
	// ...
}

Reply:
It has its pros and cons.  If a user mistakenly typed "h" instead of "s", this could really cause a problem.  If they used a value that was only a small duration of time away from the max timeout, it would be fine.

Reply:
I incorporated this suggestion, thanks.

Reply:
> It has its pros and cons. If a user mistakenly typed "h" instead of "s", this could really cause a problem. If they used a value that was only a small duration of time away from the max timeout, it would be fine.

I meant, rather than having `ParseDurationWithErrors` (which is now named `ParseHAProxyDuration` with your latest changes) return an overflow error and then having `clipHAProxyTimeoutValue` handle the overflow error by using the max value, why not simply have the parse function return the max value in case of overflows?  If you prefer returning an overflow error and handling it in `clipHAProxyTimeoutValue`, that's fine, I was just thinking you could simplify the logic a little.

Reply:
My intention was that it would be more easily reused if we left it up to the caller to interpret the error and choose what to do with it.